diff --git a/temp_Qwen3NextDynamicCache.py b/temp_Qwen3_5MoeDynamicCache.py
index a144d617..df7a3d16 100644
--- a/temp_Qwen3NextDynamicCache.py
+++ b/temp_Qwen3_5MoeDynamicCache.py
@@ -1,4 +1,4 @@
-class Qwen3NextDynamicCache:
+class Qwen3_5MoeDynamicCache:
     """
     A dynamic cache that can handle both the attention cache (which has a seq_len dimension) and the linear attention
     cache (which has a constant shape regardless of seq_len).
@@ -14,7 +14,7 @@ class Qwen3NextDynamicCache:
 
     is_compileable = False
 
-    def __init__(self, config: Qwen3NextConfig):
+    def __init__(self, config: Qwen3_5MoeConfig):
         super().__init__()
         self.layer_types = config.layer_types
         self.transformer_layers = [
