diff --git a/temp_Qwen3NextForCausalLM.py b/temp_Qwen3_5MoeForCausalLM.py
index e8b0ea68..a8bb97b8 100644
--- a/temp_Qwen3NextForCausalLM.py
+++ b/temp_Qwen3_5MoeForCausalLM.py
@@ -1,11 +1,13 @@
-class Qwen3NextForCausalLM(Qwen3NextPreTrainedModel, GenerationMixin):
+class Qwen3_5MoeForCausalLM(Qwen3_5MoePreTrainedModel, GenerationMixin):
     _tied_weights_keys = {"lm_head.weight": "model.embed_tokens.weight"}
     _tp_plan = {"lm_head": "colwise_gather_output"}
     _pp_plan = {"lm_head": (["hidden_states"], ["logits"])}
+    config: Qwen3_5MoeTextConfig
+    _keys_to_ignore_on_load_unexpected = [r"^mtp.*", r"^model.visual.*"]
 
     def __init__(self, config):
         super().__init__(config)
-        self.model = Qwen3NextModel(config)
+        self.model = Qwen3_5MoeTextModel(config)
         self.vocab_size = config.vocab_size
         self.lm_head = nn.Linear(config.hidden_size, config.vocab_size, bias=False)
         self.router_aux_loss_coef = config.router_aux_loss_coef
@@ -22,7 +24,7 @@ class Qwen3NextForCausalLM(Qwen3NextPreTrainedModel, GenerationMixin):
         input_ids: torch.LongTensor | None = None,
         attention_mask: torch.Tensor | None = None,
         position_ids: torch.LongTensor | None = None,
-        past_key_values: Qwen3NextDynamicCache | None = None,
+        past_key_values: Qwen3_5MoeDynamicCache | None = None,
         inputs_embeds: torch.FloatTensor | None = None,
         labels: torch.LongTensor | None = None,
         use_cache: bool | None = None,
@@ -40,9 +42,9 @@ class Qwen3NextForCausalLM(Qwen3NextPreTrainedModel, GenerationMixin):
         Example:
 
         ```python
-        >>> from transformers import AutoTokenizer, Qwen3NextForCausalLM
+        >>> from transformers import AutoTokenizer, Qwen3_5MoeForCausalLM
 
-        >>> model = Qwen3NextForCausalLM.from_pretrained("Qwen/Qwen3-Next-80B-A3B-Instruct")
+        >>> model = Qwen3_5MoeForCausalLM.from_pretrained("Qwen/Qwen3-Next-80B-A3B-Instruct")
         >>> tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen3-Next-80B-A3B-Instruct")
 
         >>> prompt = "Hey, are you conscious? Can you talk to me?"
