diff --git a/temp_Qwen3NextMLP.py b/temp_Qwen3_5MoeMLP.py
index 44cc4135..577a5b8d 100644
--- a/temp_Qwen3NextMLP.py
+++ b/temp_Qwen3_5MoeMLP.py
@@ -1,9 +1,9 @@
-class Qwen3NextMLP(nn.Module):
-    def __init__(self, config, intermediate_size=None):
+class Qwen3_5MoeMLP(nn.Module):
+    def __init__(self, config: Qwen3_5MoeConfig, intermediate_size: int):
         super().__init__()
         self.config = config
         self.hidden_size = config.hidden_size
-        self.intermediate_size = config.intermediate_size if intermediate_size is None else intermediate_size
+        self.intermediate_size = intermediate_size
         self.gate_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
         self.up_proj = nn.Linear(self.hidden_size, self.intermediate_size, bias=False)
         self.down_proj = nn.Linear(self.intermediate_size, self.hidden_size, bias=False)
